src/persist-client/src/read.rs

```diff
@@ -62,6 +62,126 @@ impl ReaderId {
     }
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(bound(
+    serialize = "T: Timestamp + Codec64",
+    deserialize = "T: Timestamp + Codec64"
+))]
+#[serde(
+    into = "SerdeReadEnrichedHollowBatch",
+    from = "SerdeReadEnrichedHollowBatch"
+)]
+/// A `HollowBatch` plus the metadata necessary to apply the appropriate read
+/// style's semantics (i.e. snapshotting or listening).
+pub struct ReaderEnrichedHollowBatch<T> {
+    pub(crate) shard_id: ShardId,
+    pub(crate) as_of: Antichain<T>,
+    pub(crate) frontier: Option<Antichain<T>>,
+    pub(crate) batch: HollowBatch<T>,
+}
+
+impl<T> ReaderEnrichedHollowBatch<T>
+where
+    T: Timestamp + Lattice + Codec64,
+{
+    /// Signals whether or not `self` should downgrade the `Capability` its
+    /// presented alongside.
+    pub fn generate_progress(&self) -> Option<Antichain<T>> {
+        // As-is only `Listen`-style readers have a `frontier` of `Some`. (TODO:
+        // make this more well structured)
+        match self.frontier {
+            Some(_) => Some(self.batch.desc.upper().clone()),
+            None => None,
+        }
+    }
+}
+
+/// Fetches and filters the embedded `HollowBatch`.
+///
+/// See [ReadHandle::hollow_batch] for details.
+#[derive(Debug)]
+pub struct HollowBatchFetcher<K, V, T, D> {
+    metrics: Arc<Metrics>,
+    shard_id: ShardId,
+    as_of: Antichain<T>,
+    frontier: Option<Antichain<T>>,
+    batch: HollowBatch<T>,
+    blob: Arc<dyn Blob + Send + Sync>,
+    _phantom: PhantomData<(K, V, T, D)>,
+}
+
+impl<K, V, T, D> HollowBatchFetcher<K, V, T, D>
+where
+    K: Debug + Codec,
+    V: Debug + Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Semigroup + Codec64,
+{
+    /// The frontier at which we're outputting the contents of the shard.
+    pub fn as_of(&self) -> &Antichain<T> {
+        &self.as_of
+    }
+
+    /// Fetches a `HollowBatch` from S3 and applies the appropriate
+    /// `as_of`/`frontier` filtering.
+    #[instrument(level = "debug", name = "snap::next", skip_all, fields(shard = %self.shard_id))]
+    pub async fn fetch(&mut self) -> Vec<ListenEvent<K, V, T, D>> {
```

Comment:
i've been imagining that there would be a `fetch_batch` sort of method on ReadHandle that takes in an enriched hollow batch and directly hands back updates and frontier information. but maybe i'm missing something! what does the additional `HollowBatchFetcher` and `fetch` call get us?

Reply:
Can get reduced away, I'm almost certain.